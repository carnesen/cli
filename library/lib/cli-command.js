"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliCommand = void 0;
const constants_1 = require("./constants");
/**
 *
 * A factory function for creating CLI command commands
 * @param options
 * @returns The newly-created `command`.
 */
function CliCommand(options) {
    return {
        ...options,
        commandType: constants_1.CLI_COMMAND,
    };
}
exports.CliCommand = CliCommand;
//# sourceMappingURL=cli-command.js.map