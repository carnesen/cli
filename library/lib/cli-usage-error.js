"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliUsageError = exports.CLI_USAGE_ERROR = void 0;
exports.CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';
/**
 * A custom error with code {@linkcode CLI_USAGE_ERROR}
 * @remarks
 * {@linkcode runCliAndExit} catches this class of error and prints command-line usage
 * (i.e. "Usage: cloud users lists ...") for the current command. If the `message`
 * property is provided, `runCliAndExit` will normal `Error`. See also print `Error: <your
 * message>`. See also {@linkcode CliTerseError}.
 * @example
 * ```typescript
 * throw new CliUsageError("Expected one or more values")
 * ```
 */
class CliUsageError extends Error {
    /**
     *
     * @param message An `Error` message
     * @param commandStack Used internally for argument processing
     */
    constructor(message, commandStack) {
        super(message);
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "commandStack", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.code = exports.CLI_USAGE_ERROR;
        this.commandStack = commandStack;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.CliUsageError = CliUsageError;
//# sourceMappingURL=cli-usage-error.js.map