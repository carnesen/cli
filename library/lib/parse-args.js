"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseArgs = void 0;
const cli_usage_error_1 = require("./cli-usage-error");
/**
 * Calls the parse method of an ArgParser
 *
 * @param argParser - An ArgParser
 * @param args - An array of string arguments
 * @param separator - A string that will be prepended to error messages (e.g. "--users")
 * @returns The result of parse
 */
async function parseArgs(argParser, args, separator, commandStack) {
    const { required, placeholder, parse } = argParser;
    const fullContext = [separator, placeholder].filter((str) => Boolean(str)).join(' ');
    const prefix = fullContext ? `${fullContext} : ` : '';
    if (required && (!args || args.length === 0)) {
        throw new cli_usage_error_1.CliUsageError(`${prefix}argument is required`, commandStack);
    }
    try {
        return await parse(args);
    }
    catch (exception) {
        if (exception && typeof exception.message === 'string') {
            exception.message = `${prefix}${exception.message}`;
        }
        if (exception && exception.code === cli_usage_error_1.CLI_USAGE_ERROR) {
            exception.commandStack = commandStack;
        }
        throw exception;
    }
}
exports.parseArgs = parseArgs;
//# sourceMappingURL=parse-args.js.map