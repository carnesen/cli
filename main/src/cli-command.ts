import { CLI_COMMAND } from './constants';
import {
  AnyArgParser,
  AnyNamedArgParsers,
  ValueFromArgParser,
  NamedValues,
  ICliArgParser,
} from './cli-arg-parser';

export type CliCommandOptions<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> = {
  /**
   * Name used in command-line usage
   */
  name: string;
  /**
   * The function that defines your command logic.
   *
   * @param positionalValue
   * @param namedValues
   * @param escapedValue
   * @returns A value (synchronously) or a `Promise` that resolves to a value
   * @remarks
   * By default, {@linkcode runCliAndExit} `console.log`s the value returned by `action`.
   * The type of the `args` argument received by `action` is derived by the `args`
   * property of the command. Similarly, the `options` argument type is derived from
   * `command.options`.
   *
   */
  action: (
    positionalValue: ValueFromArgParser<TPositionalArgParser>,
    namedValues: NamedValues<TNamedArgParsers>,
    escapedValue: ValueFromArgParser<TEscapedArgParser>,
  ) => any;
  positionalArgParser?: TPositionalArgParser;
  namedArgParsers?: TNamedArgParsers;
  escapedArgParser?: TEscapedArgParser;
  /**
   * A string that will be included in `Usage:` if present.
   */
  description?: string;
  /**
   * If true, don't show this command in usage docs.
   */
  hidden?: boolean;
};

/**
 * Interface describing a "command" command
 * @typeParam TPositionalArgParser Type of the "positional" arguments parser
 * @typeParam TNamedArgParsers Type of the ""
 */
export interface ICliCommand<
  TPositionalArgParser extends AnyArgParser,
  TNamedArgParsers extends AnyNamedArgParsers,
  TEscapedArgParser extends AnyArgParser
> extends CliCommandOptions<TPositionalArgParser, TNamedArgParsers, TEscapedArgParser> {
  commandType: typeof CLI_COMMAND;
}

/**
 *
 * A factory function for creating CLI command commands
 * @param options
 * @returns The newly-created `command`.
 */
export function CliCommand<
  TPositional extends AnyArgParser = ICliArgParser<undefined, false>,
  TNamed extends AnyNamedArgParsers = any,
  TEscaped extends AnyArgParser = ICliArgParser<undefined, false>
>(
  options: CliCommandOptions<TPositional, TNamed, TEscaped>,
): ICliCommand<TPositional, TNamed, TEscaped> {
  return {
    ...options,
    commandType: CLI_COMMAND,
  };
}
