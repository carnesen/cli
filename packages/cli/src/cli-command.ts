import { ValueFromCliArgGroup, CliArgGroup } from './cli-arg-group';
import { CliLogger } from './cli-logger';
import { CliColor } from './cli-color';
import { AnyCliDescription } from './cli-description';

/** "kind" of a [[`ICliCommand`]] */
export const CLI_COMMAND = 'CLI_COMMAND';

/** Options for [[`CliCommand`]] */
export type CliCommandOptions<
	PositionalArgGroup extends CliArgGroup = CliArgGroup,
	NamedArgGroups extends {
		[name: string]: CliArgGroup;
	} = {
		[name: string]: CliArgGroup;
	},
	DoubleDashArgGroup extends CliArgGroup = CliArgGroup,
> = {
	/** Identifier for this command in command-line usage */
	name: string;

	/** Function or async function that implements the command */
	action: (input: {
		/** @deprecated Use `color` instead */
		ansi: CliColor;
		color: CliColor;
		/** @deprecated Use `logger` instead */
		console: CliLogger;
		doubleDashValue: ValueFromCliArgGroup<DoubleDashArgGroup>;
		logger: CliLogger;
		namedValues: {
			[K in keyof NamedArgGroups]: ValueFromCliArgGroup<NamedArgGroups[K]>;
		};
		positionalValue: ValueFromCliArgGroup<PositionalArgGroup>;
	}) => any;

	/** A [[`CliArgGroup`]] for the arguments before the first separator argument */
	positionalArgGroup?: PositionalArgGroup;

	/** A [[`CliArgGroup`]] for the arguments passed as "--name value" */
	namedArgGroups?: NamedArgGroups;

	/** A [[`CliArgGroup`]] for the arguments after a lone "--" */
	doubleDashArgGroup?: DoubleDashArgGroup;

	/** A sentence or two about this command for command-line usage */
	description?: AnyCliDescription;

	/** If `true`, don't show this command in command-line usage */
	hidden?: boolean;
};

/** An object that defines a CLI command and its arguments */
export interface ICliCommand<
	PositionalArgGroup extends CliArgGroup = CliArgGroup,
	NamedArgGroups extends {
		[name: string]: CliArgGroup;
	} = {
		[name: string]: CliArgGroup;
	},
	DoubleDashArgGroup extends CliArgGroup = CliArgGroup,
> extends CliCommandOptions<
		PositionalArgGroup,
		NamedArgGroups,
		DoubleDashArgGroup
	> {
	/** The string literal [[`CLI_COMMAND`]] */
	kind: typeof CLI_COMMAND;
}

/** A factory for [[`ICliCommand`]]s */
export function CliCommand<
	PositionalArgGroup extends CliArgGroup = CliArgGroup<unknown, false>,
	NamedArgGroups extends {
		[name: string]: CliArgGroup;
	} = {
		[name: string]: CliArgGroup;
	},
	DoubleDashArgGroup extends CliArgGroup = CliArgGroup<unknown, false>,
>(
	options: CliCommandOptions<
		PositionalArgGroup,
		NamedArgGroups,
		DoubleDashArgGroup
	>,
): ICliCommand<PositionalArgGroup, NamedArgGroups, DoubleDashArgGroup> {
	return {
		...options,
		kind: CLI_COMMAND,
	};
}
