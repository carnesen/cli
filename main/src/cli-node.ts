import { AnyParser, AnyNamedParsers } from './cli-arg-parser';
import { ICliBranch } from './cli-branch';
import { ICliCommand } from './cli-command';

type AnyCliCommand = ICliCommand<AnyParser, AnyNamedParsers, AnyParser>;
export type BranchOrCommand = ICliBranch | AnyCliCommand;

export type Node = {
  parents: ICliBranch[];
  current: BranchOrCommand;
};

export type Leaf = { parents: ICliBranch[]; current: AnyCliCommand };
