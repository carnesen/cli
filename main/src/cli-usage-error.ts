import { Node } from './cli-node';

/** "code" of a {@link CliUsageError} */
export const CLI_USAGE_ERROR = 'CLI_USAGE_ERROR';

/**
 * Thrown to print command usage and an error message but not a `stack`
 */
export class CliUsageError extends Error {
  /** The string constant {@linkcode CLI_USAGE_ERROR} */
  public readonly code: typeof CLI_USAGE_ERROR;

  /** Used internally for constructing the command-line usage string */
  public node?: Node;

  /**
   * @param message If provided, {@linkcode runCliAndExit} will also print "Error: \<your
   * message\>"
   * @param node Used internally for constructing the command-line usage string
   */
  constructor(message?: string, node?: Node) {
    super(message);
    this.code = CLI_USAGE_ERROR;
    this.node = node;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
