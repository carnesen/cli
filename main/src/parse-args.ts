import { Leaf } from './cli-node';
import { CliUsageError, CLI_USAGE_ERROR } from './cli-usage-error';
import { AnyArgParser } from './cli-arg-parser';

/**
 * Calls the parse method of an ArgParser
 *
 * @param argParser - An ArgParser
 * @param args - An array of string arguments
 * @param separator - A string that will be prepended to error messages (e.g. "--users")
 * @returns The result of parse
 */
export async function parseArgs(
  argParser: AnyArgParser,
  args: string[] | undefined,
  separator: string | undefined,
  locationInCommandTree: Leaf,
): Promise<any> {
  const { required, placeholder, parse } = argParser;
  const fullContext = [separator, placeholder].filter((str) => Boolean(str)).join(' ');
  const prefix = fullContext ? `${fullContext} : ` : '';
  if (required && (!args || args.length === 0)) {
    throw new CliUsageError(`${prefix}argument is required`, locationInCommandTree);
  }
  try {
    return await parse(args);
  } catch (exception) {
    if (exception && typeof exception.message === 'string') {
      exception.message = `${prefix}${exception.message}`;
    }
    if (exception && exception.code === CLI_USAGE_ERROR) {
      exception.locationInCommandTree = locationInCommandTree;
    }
    throw exception;
  }
}
