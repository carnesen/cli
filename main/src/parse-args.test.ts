import { runAndCatch } from '@carnesen/run-and-catch';

import { parseArgs } from './parse-args';
import { CLI_USAGE_ERROR } from './cli-usage-error';
import {
	dummyValuedParser,
	dummyRequiredValuedParser,
	DUMMY_ARG_PARSER_THROWN_INTENTIONALLY,
	DUMMY_ARG_PARSER_THROW,
	DUMMY_ARG_PARSER_THROW_NON_TRUTHY,
} from './dummy-arg-parsers-for-testing';
import { CliCommand } from './cli-command';
import { CliCommandNode } from './cli-node';

const node: CliCommandNode = {
	current: CliCommand({ name: 'foo', action() {} }),
	parents: [],
};

describe(parseArgs.name, () => {
	it(`returns parse(args) if an args with length >= 1 is passed`, async () => {
		const args = ['foo'];
		expect(await parseArgs(dummyValuedParser, args, undefined, node)).toBe(
			dummyValuedParser.parse(args),
		);
		expect(
			await parseArgs(dummyRequiredValuedParser, args, undefined, node),
		).toBe(dummyRequiredValuedParser.parse(args));
	});

	it(`if not required, returns parse(args) if args is an empty array or undefined`, async () => {
		expect(await parseArgs(dummyValuedParser, [], undefined, node)).toBe(
			dummyValuedParser.parse([]),
		);
		expect(await parseArgs(dummyValuedParser, undefined, undefined, node)).toBe(
			dummyValuedParser.parse(undefined),
		);
	});

	it(`if required, throws usage error "argument is required" if args is an empty array or undefined`, async () => {
		for (const args of [undefined, [] as string[]]) {
			const exception = await runAndCatch(
				parseArgs,
				dummyRequiredValuedParser,
				args,
				undefined,
				node,
			);
			expect(exception.code).toBe(CLI_USAGE_ERROR);
			expect(exception.message).toMatch(/argument is required/i);
			expect(exception.message).toMatch(dummyRequiredValuedParser.placeholder);
		}
	});

	it(`if throws "argument is required", expect message to match snapshot`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyRequiredValuedParser,
			undefined,
			undefined,
			node,
		);
		expect(exception.message).toMatch('argument is required');
		expect(exception.message).toMatchSnapshot();
	});

	it(`if throws "argument is required" with context, expect message to match snapshot`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyRequiredValuedParser,
			undefined,
			'context',
			node,
		);
		expect(exception.message).toMatch('argument is required');
		expect(exception.message).toMatchSnapshot();
	});

	it(`throws if parse does with a context/placeholder enhanced message`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyValuedParser,
			[DUMMY_ARG_PARSER_THROW],
			undefined,
			node,
		);
		expect(exception.message).toMatch(DUMMY_ARG_PARSER_THROWN_INTENTIONALLY);
		expect(exception.message).toMatch(dummyValuedParser.placeholder);
		expect(exception.message).toMatchSnapshot();
	});

	it(`just re-throws exception if parse throws a non-truthy exception`, async () => {
		const exception = await runAndCatch(
			parseArgs,
			dummyValuedParser,
			[DUMMY_ARG_PARSER_THROW_NON_TRUTHY],
			undefined,
			node,
		);
		expect(exception).not.toBeTruthy();
	});
});
