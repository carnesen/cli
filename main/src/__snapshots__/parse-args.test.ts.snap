// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseArgs if throws "parser is required" with context, expect message to match snapshot 1`] = `"context <foo> : argument is required"`;

exports[`parseArgs if throws "parser is required", expect message to match snapshot 1`] = `"<foo> : argument is required"`;

exports[`parseArgs throws if parse does with a context/placeholder enhanced message 1`] = `"thrown intentionally"`;
